exit/test_exit.c

```diff
@@ -0,0 +1,513 @@
+	/* child */
+	if (pid == 0) {
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int ret;
+
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+		/* Try to kill process which exited */
+		errno = 0;
+		ret = kill(pid, SIGKILL);
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_EQUAL_INT(ESRCH, errno);
+	}
+}
+
+
+TEST(_exit, close_streams)
+{
+	/*
+	 * Open pipe, when child exits one side of pipe should be closed,
+	 * so that call to write should fail.
+	 */
+	pid_t pid;
+	int pipefd[2];
+	int status;
+
+	TEST_ASSERT_EQUAL_INT(0, pipe(pipefd));
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int ret;
+
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+		close(pipefd[0]); /* close read pipe end */
+		errno = 0;
+		ret = write(pipefd[1], TEST_STR, sizeof(TEST_STR));
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_EQUAL_INT(EPIPE, errno);
+		close(pipefd[1]); /* close write pipe end */
+	}
+}
+
+
+TEST(_exit, orphaned_child)
+{
+	/* Test if parent _exit affect child process */
+	pid_t pid;
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	if (pid == 0) {
+		pid = fork();
+		TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+		/* grandchild */
+		if (pid == 0) {
+			/* Sleep for a while to ensure that parent already exited */
+			sleep(1);
+			creat(PATH3, DEFFILEMODE);
+			_exit(EXIT_SUCCESS);
+		}
+		/* parent */
+		else {
+			/* parent exits right away */
+			_exit(EXIT_SUCCESS);
+		}
+	}
+	/* grandparent */
+	else {
+		int status;
+
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+
+		/* Check if file exists, if so child sleep was interrupted */
+		TEST_ASSERT_EQUAL_INT(-1, access(PATH3, F_OK));
+
+		sleep(2);
+		/* At this moment if child wasn't affected file should exist */
+		TEST_ASSERT_EQUAL_INT(0, access(PATH3, F_OK));
+
+		remove(PATH3);
+	}
+}
+
+
+TEST(_exit, new_parent_id)
+{
+	/* Test that child acquire new parent ID */
+	pid_t pid;
+	int pipefd[2];
+
+	/* pipe needed for communication between grandparent and grandchild (can't use asserts in child) */
+	TEST_ASSERT_EQUAL_INT(0, pipe(pipefd));
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	if (pid == 0) {
+		pid = fork();
+		TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+		/* grandchild */
+		if (pid == 0) {
+			pid_t old_ppid, new_ppid;
+
+			close(pipefd[0]);
+			old_ppid = getppid();
+			/* Sleep until parents exits */
+			sleep(2);
+			/* Parent process id should be 1 (init pid) */
+			new_ppid = getppid();
+			write(pipefd[1], &old_ppid, sizeof(pid_t));
+			write(pipefd[1], &new_ppid, sizeof(pid_t));
+			close(pipefd[1]);
+			_exit(EXIT_SUCCESS);
+		}
+		/* parent */
+		else {
+			close(pipefd[0]);
+			close(pipefd[1]);
+			/* Sleep until child gets his parent ID */
+			sleep(1);
+			_exit(EXIT_SUCCESS);
+		}
+	}
+	/* grandparent */
+	else {
+		pid_t old_ppid, new_ppid;
+		int status;
```

Comment:
variable declaration deep in test body, not in the beginning. Also names do not follow coding convention.

Reply:
I think this approach is more readable and was used in other files in the past: https://github.com/phoenix-rtos/phoenix-rtos-tests/blob/4334a5c63a0d66c7b884f34259c76b60ccdd50d5/socket/test_socket.c?plain=1#L623-L670

Reply:
Yes, this was used in the past. This code is pushed now and it should conform to current coding convention.

Reply:
@adamdebek still the variable names are not correct, please use camelCase.